File: src/Sources/IGN.php
<?php

/**
 * AUXIO - Fuente IGN: actividad sísmica via GeoRSS
 */

require_once __DIR__ . '/../Alert.php';

class IGNSource {
    private const FEED_URL = "https://www.ign.es/ign/RssTools/sismologia.xml";

    private const SEV_THRESHOLDS = [
        5.5 => "red",
        4.0 => "orange",
        2.5 => "yellow",
    ];

    /**
     * Map de magnitud a severidad (escala Richter)
     * < 2.5  : green   — generalmente no sentido
     * 2.5–4.0: yellow  — sentido, daños menores
     * 4.0–5.5: orange  — daños moderados
     * >= 5.5 : red     — daños significativos
     */
    private static function severityFromMagnitude(float $mag): string {
        if ($mag >= 5.5) return "red";
        if ($mag >= 4.0) return "orange";
        if ($mag >= 2.5) return "yellow";
        return "green";
    }

    /**
     * Descargar feed GeoRSS del IGN
     */
    private static function fetchFeed(): string {
        $context = stream_context_create([
            'http' => [
                'method' => 'GET',
                'header' => "User-Agent: AUXIO/1.0\r\n",
                'timeout' => 30,
            ]
        ]);

        $response = @file_get_contents(self::FEED_URL, false, $context);
        if ($response === false) {
            throw new Exception("Error fetching IGN feed");
        }
        return $response;
    }

    /**
     * Extraer magnitud desde descripción (regex)
     * Ejemplo: "Se ha producido un terremoto de magnitud 3.1 en..."
     */
    private static function extractMagnitude(string $description): ?float {
        if (preg_match('/magnitud\s+([\d.]+)/', $description, $matches)) {
            return (float)$matches[1];
        }
        return null;
    }

    /**
     * Extraer región desde descripción
     */
    private static function extractRegion(string $description): ?string {
        if (preg_match('/magnitud\s+[\d.]+\s+en\s+(.+?)\s+en la fecha/', $description, $matches)) {
            return trim($matches[1]);
        }
        return null;
    }

    /**
     * Extraer fecha desde descripción
     */
    private static function extractDate(string $description): ?string {
        if (preg_match('/en la fecha\s+(\d{2}\/\d{2}\/\d{4}\s+\d{1,2}:\d{2}:\d{2})/', $description, $matches)) {
            return $matches[1];
        }
        return null;
    }

    /**
     * Parsear feed GeoRSS en Alerts
     */
    private static function parseFeed(string $xml): array {
        $alerts = [];

        try {
            $root = new SimpleXMLElement($xml);
        } catch (Exception $e) {
            throw new Exception("Invalid RSS XML: {$e->getMessage()}");
        }

        $channel = $root->channel ?? $root;

        foreach ($channel->item ?? [] as $item) {
            $description = '';
            if (isset($item->description)) {
                $description = trim((string)$item->description);
            }

            // Extraer magnitud
            $magnitude = self::extractMagnitude($description);
            if ($magnitude === null) {
                continue;  // Skip sin magnitud
            }

            // Extraer región
            $region = self::extractRegion($description);

            // Extraer fecha
            $onset = self::extractDate($description);

            // Extraer coordenadas (GeoRSS)
            $coords = '';
            $namespaces = $item->getNamespaces(true);
            if (isset($namespaces['geo'])) {
                $geoNS = $item->children($namespaces['geo']);
                if (isset($geoNS->lat) && isset($geoNS->long)) {
                    $lat = trim((string)$geoNS->lat);
                    $lon = trim((string)$geoNS->long);
                    if ($lat && $lon) {
                        $coords = " ({$lat}, {$lon})";
                    }
                }
            }

            // Link a página de detalle
            $web = null;
            if (isset($item->link)) {
                $web = trim((string)$item->link);
            }

            $severity = self::severityFromMagnitude($magnitude);
            $headline = "Terremoto M{$magnitude}";
            if ($region) {
                $headline .= " en {$region}";
            }

            $fullDescription = "Magnitud {$magnitude}";
            if ($region) {
                $fullDescription .= " en {$region}";
            }
            if ($onset) {
                $fullDescription .= ", {$onset}";
            }
            if ($coords) {
                $fullDescription .= $coords;
            }

            $alerts[] = new Alert(
                source: 'ign',
                severity: $severity,
                headline: $headline,
                description: $fullDescription,
                area: $region,
                event_type: 'Terremoto',
                onset: $onset,
                sender: 'Instituto Geográfico Nacional',
                web: $web,
            );
        }

        return $alerts;
    }

    /**
     * Fetch: obtener actividad sísmica del IGN
     */
    public static function fetch(): array {
        try {
            $xml = self::fetchFeed();
            return self::parseFeed($xml);
        } catch (Exception $exc) {
            echo "[ign] Error fetching seismic data: {$exc->getMessage()}\n";
            return [];
        }
    }
}

File: sources/ign.py
"""Fuente IGN: actividad sismica via GeoRSS del Instituto Geografico Nacional."""

import re
import xml.etree.ElementTree as ET
from typing import List
from urllib.request import Request, urlopen
from urllib.error import URLError

from schema import Alert

FEED_URL = "https://www.ign.es/ign/RssTools/sismologia.xml"
GEO_NS = "{http://www.w3.org/2003/01/geo/wgs84_pos#}"

# Magnitude-to-severity mapping (Richter-like scale thresholds).
# < 2.5  : green  — generalmente no sentido
# 2.5–4.0: yellow — sentido, danos menores
# 4.0–5.5: orange — danos moderados
# >= 5.5 : red    — danos significativos
_SEV_THRESHOLDS = [(5.5, "red"), (4.0, "orange"), (2.5, "yellow")]

# Regex to extract magnitude and region from the description text.
# Example: "Se ha producido un terremoto de magnitud 3.1 en SW EL PINAR
#           DE EL HIERRO.IHI en la fecha 18/03/2019 19:34:55 en la
#           siguiente localizacion: 27.62,-18.0859"
_RE_MAGNITUDE = re.compile(r"magnitud\s+([\d.]+)")
_RE_REGION = re.compile(r"magnitud\s+[\d.]+\s+en\s+(.+?)\s+en la fecha")
_RE_DATE = re.compile(r"en la fecha\s+(\d{2}/\d{2}/\d{4}\s+\d{1,2}:\d{2}:\d{2})")


def _severity_from_magnitude(mag: float) -> str:
    """Map a numeric magnitude to a severity level."""
    for threshold, level in _SEV_THRESHOLDS:
        if mag >= threshold:
            return level
    return "green"


def _fetch_feed() -> bytes:
    """Download the IGN GeoRSS feed."""
    req = Request(FEED_URL, headers={"User-Agent": "AUXIO/1.0"})
    with urlopen(req, timeout=30) as resp:
        return resp.read()


def _parse_feed(xml_bytes: bytes) -> List[Alert]:
    """Parse the GeoRSS XML into a list of Alert objects."""
    alerts: List[Alert] = []
    root = ET.fromstring(xml_bytes)

    channel = root.find("channel")
    if channel is None:
        channel = root  # fallback: items at root level

    for item in channel.findall("item"):
        description = ""
        desc_el = item.find("description")
        if desc_el is not None and desc_el.text:
            description = desc_el.text.strip()

        # Extract magnitude
        mag_match = _RE_MAGNITUDE.search(description)
        if not mag_match:
            continue  # skip items without magnitude info
        magnitude = float(mag_match.group(1))

        # Extract region
        region_match = _RE_REGION.search(description)
        region = region_match.group(1).strip() if region_match else None

        # Extract date/time
        date_match = _RE_DATE.search(description)
        onset = date_match.group(1) if date_match else None

        # Extract coordinates
        lat_el = item.find(f"{GEO_NS}lat")
        lon_el = item.find(f"{GEO_NS}long")
        coords = ""
        if lat_el is not None and lon_el is not None:
            lat_text = lat_el.text.strip() if lat_el.text else ""
            lon_text = lon_el.text.strip() if lon_el.text else ""
            if lat_text and lon_text:
                coords = f" ({lat_text}, {lon_text})"

        # Link to detail page
        link_el = item.find("link")
        web = None
        if link_el is not None and link_el.text:
            web = link_el.text.strip()

        severity = _severity_from_magnitude(magnitude)
        headline = f"Terremoto M{magnitude}"
        if region:
            headline += f" en {region}"

        full_description = f"Magnitud {magnitude}"
        if region:
            full_description += f" en {region}"
        if onset:
            full_description += f", {onset}"
        if coords:
            full_description += coords

        alerts.append(
            Alert(
                source="ign",
                severity=severity,
                headline=headline,
                description=full_description,
                area=region,
                event_type="Terremoto",
                onset=onset,
                sender="Instituto Geografico Nacional",
                web=web,
            )
        )

    return alerts


def fetch() -> List[Alert]:
    """Fetch current seismic activity from IGN GeoRSS feed.

    Returns a list of Alert objects following the AUXIO schema.
    Returns an empty list if the request fails.
    """
    try:
        xml_bytes = _fetch_feed()
        return _parse_feed(xml_bytes)
    except (URLError, ET.ParseError, ValueError) as exc:
        print(f"[ign] Error fetching seismic data: {exc}")
        return []
